tests/test_dynamo_db_block.py

```diff
@@ -0,0 +1,213 @@
+from time import sleep
+from nio.common.signal.base import Signal
+from nio.util.support.block_test_case import NIOBlockTestCase
+from nio.modules.threading import spawn
+from ..dynamo_db_block import DynamoDB
+from boto.exception import JSONResponseError
+from boto.dynamodb2.fields import HashKey, RangeKey
+
+
+class SaveCounterDynamoDB(DynamoDB):
+
+    def __init__(self):
+        super().__init__()
+        self._count = 0
+
+    def _save_signals_to_table(self, table, signals):
+        self._count += 1
+        super()._save_signals_to_table(table, signals)
+
+class TestDynamo(NIOBlockTestCase):
+
+    @patch(DynamoDB.__module__ + '.connect_to_region')
+    def test_connect(self, connect_func):
+        """ Make sure we connect with the right creds """
+        blk = DynamoDB()
+        self.configure_block(blk, {
+            'region': 'us_east_1',
+            'creds': {
+                'access_key': 'FAKEKEY',
+                'access_secret': 'FAKESECRET'
+            }
+        })
+        connect_func.assert_called_once_with(
+            'us-east-1',
+            aws_access_key_id='FAKEKEY',
+            aws_secret_access_key='FAKESECRET')
+
+    @patch(DynamoDB.__module__ + '.Table.create')
+    @patch('boto.dynamodb2.connect_to_region')
+    @patch('boto.dynamodb2.table.Table.count')
+    def test_no_table(self, count_func, connect_func, create_func):
+        """ Assert that tables that aren't found are created """
+
+        # Spit out the resource not found exception we get from AWS
+        count_func.side_effect = [5, JSONResponseError(
+            400,
+            "{'message': 'Requested resource not found: Table: T notfound', "
+            "'__type': 'com.amazonaws.dynamodb.v20120810"
+            "#ResourceNotFoundException'}")]
+        blk = DynamoDB()
+        self.configure_block(blk, {
+            'log_level': 'DEBUG'
+        })
+
+        # Make sure boto didn't create the table that was found
+        blk._get_table('should_be_found')
+        self.assertEqual(count_func.call_count, 1)
+        self.assertEqual(create_func.call_count, 0)
+
+        # Make sure boto tried to create the table
+        blk._get_table('should_not_be_found')
+        self.assertEqual(count_func.call_count, 2)
+        self.assertEqual(create_func.call_count, 1)
+
+        # Make sure the two tables are cached now
+        self.assertIn('should_be_found', blk._table_cache)
+        self.assertIn('should_not_be_found', blk._table_cache)
+
+        # Subsequent get table calls should hit the cache, we shouldn't see
+        # the create function or the count function get called again
+        blk._get_table('should_be_found')
+        blk._get_table('should_not_be_found')
+        self.assertEqual(count_func.call_count, 2)
+        self.assertEqual(create_func.call_count, 1)
+
+    @patch('boto.dynamodb2.table.BatchTable.put_item')
+    @patch('boto.dynamodb2.table.Table.count')
+    @patch(DynamoDB.__module__ + '.connect_to_region')
```

Comment:
Ok, I added a commit (aed990f) that I think makes it more consistent. After reading [this](https://docs.python.org/3/library/unittest.mock.html#where-to-patch) I think we should pretty much always be patching from the block class since that is where the code will be looked up. The only exception is the `BatchTable` since that actually isn't imported in the class and gets looked up dynamically at runtime.